python/tests/ibd.py

```diff
@@ -0,0 +1,421 @@
+        if sample0 == sample1:
+            raise ValueError("Samples in pair must have different node IDs.")
+        elif sample0 > sample1:
+            sample0, sample1 = sample1, sample0
+
+        i0 = self.sample_id_map[sample0]
+        i1 = self.sample_id_map[sample1]
+
+        # Calculate the position of the sample pair in the vector.
+        index = (
+            (self.num_samples) * (self.num_samples - 1) / 2
+            - (self.num_samples - i0) * (self.num_samples - i0 - 1) / 2
+            + i1
+            - i0
+            - 1
+        )
+
+        return int(index)
+
+    def find_ibd_segments(self, min_length=0, max_time=None):
+        """
+        The wrapper for the procedure that calculates IBD segments.
+        """
+
+        # Set up an iterator over the edges in the tree sequence.
+        mygen = iter(self.ts.edges())
+        e = next(mygen)
+        nodes_to_process = []
+
+        # Iterate over the edges.
+        while e is not None:
+            # Process all edges with the same parent node.
+            self.current_parent = e.parent
+            self.current_time = self.tables.nodes.time[self.current_parent]
+            if self.max_time is not None and self.current_time > self.max_time:
+                # Stop looking for IBD segments once the
+                # processed nodes are older than the max time.
+                break
+
+            seg = Segment(e.left, e.right, e.child)
+
+            # Create a SegmentList() holding all segments that descend from seg.
+            list_to_add = SegmentList()
+            u = seg.node
+            if u in self.samples:
+                list_to_add.add(seg)
+            else:
+                if self.A[u] is not None:
+                    s = self.A[u].head
+                    while s is not None:
+                        intvl = (
+                            max(seg.left, s.left),
+                            min(seg.right, s.right),
+                        )
+                        if intvl[1] - intvl[0] > 0:
+                            list_to_add.add(Segment(intvl[0], intvl[1], s.node))
+                        s = s.next
+
+            pn = self.calculate_ibd_segs(self.current_parent, list_to_add)
+            nodes_to_process += pn
+
+            # For parents that are also samples
+            if (self.current_parent in self.samples) and self.PARENT_SHOULD_BE_ADDED:
+                singleton_seg = SegmentList()
+                singleton_seg.add(
+                    Segment(0, self.ts.sequence_length, self.current_parent)
+                )
+                pn = self.calculate_ibd_segs(self.current_parent, singleton_seg)
+                self.PARENT_SHOULD_BE_ADDED = False
+                nodes_to_process += pn
+
+            # Move to next edge.
+            e = next(mygen, None)
+
+            # Remove any processed nodes that are no longer needed.
+            if e is None or e.parent != self.current_parent:
```

Comment:
OK, so this is done each time we move on to a new parent. But what's it actually doing? It seems to me that we're only manipulating the ``self.parent_list`` array in this loop, so is it actually needed now?

Reply:
The end effect is to set ``self.A[n]`` to the empty list, so is there some other way we could achieve this?

Reply:
Let's talk about this in person --